docs: update docs-style-guide (#36281)
angular/angular: aio/content/guide/docs-style-guide.md

# Angular documentation style guide

This style guide covers the standards for writing [Angular documentation on angular.io](https://angular.io/docs).

> These standards ensure consistency in writing style, markdown conventions, and code snippets.

Author: Fixed throughout.

{@a display-whole-file}

### Displaying an entire code file

This Angular documentation style guide that you are currently reading has its own example application, located in the `content/examples/docs-style-guide` folder.

The following `<code-example>` displays the sample's `app.module.ts`.

<code-example path="docs-style-guide/src/app/app.module.ts" header="src/app/app.module.ts"></code-example>

The following markup produces that snippet:

```html
<code-example path="docs-style-guide/src/app/app.module.ts" header="src/app/app.module.ts">
</code-example>
```

The `path` attribute identifies the snippet's source file at the example app folder's location within `content/examples`.
In this example, that path is  `docs-style-guide/src/app/app.module.ts`.

The header tells the reader where to find the file.
Following convention, set the `header` attribute to the file's location within the example app's root folder.

<div class="alert is-helpful">

Unless otherwise noted, all code snippets in this page are from sample source code
located in the `content/examples/docs-style-guide` directory.

</div>

<div class="alert is-important">

> The doc tooling reports an error if the file identified in the path does not exist or is in the [`.git-ignore` file](https://github.com/angular/angular/blob/master/aio/content/examples/.gitignore).

Reviewer: First mention of "doc tooling" -- an earlier section that was taken out. Is there a place to go for info on doc tooling? I have been using the [Set up Tools wiki](https://github.com/ExpertSupport/doc-workflow/wiki/Set-up-Tools) and other wiki pages.

Author: Doc tooling refers to the build and compilation process of running the docs (that stuff that happens when you use `yarn serve-and-sync`), rather than workflow (the apps, version control, and processes you use when working). Would you have another suggestion for "that-stuff-that-happens-when-you-use-`yarn serve-and-sync`" if "tooling" is confusing? To me it is clear, but I might be used to it and we want to include new people. Looks like it could use clarification. 🙂

Regarding recommending a particular workflow (editor, package managers, local setup beyond necessity) as in the link you shared, I don't think we do that in the docs as far as I know. Whether one uses VSC, has the icon in the dock, or uses Homebrew as that wiki page advises, falls into the realm of preference rather than a requirement for running the docs. I could be wrong, and we might indeed like to include info on workflow, but it would be more suitable for another doc. And really, you don't actually need to do any of those things to contribute. You could edit in just about any code-friendly editor, install `yarn` using `curl` (or never even run the docs making `yarn` unnecessary), and still submit a successful PR. There are so many different workflows!

Reviewer: I think that doc tooling is itself an entire set of instructions involving yarn and git. There are overlaps in what we need to know as authors, and what developers need to know to get Angular up and running, and one big overlap is how to use the proper tools (mostly yarn but others) to create an example, or edit an example, for a doc. When this doc becomes available (@aikidave ?) we should link to it from this doc.

</code-example>
```

It renders with a red header and border in the browser:

<code-example path="styleguide/src/05-03/app/heroes/shared/hero-button/hero-button.component.avoid.ts" region="example" header="app/heroes/hero-button/hero-button.component.ts">
</code-example>

Alternatively, the HTML could include the `avoid` class as in the following:

```html
<code-example
  path="docs-style-guide/src/app/not-great.component.ts"
  header="docs-style-guide/src/app/not-great.component.ts"
  region="not-great"
  class="avoid">
</code-example>
```

Explicitly applying the class `avoid` causes the same result of a red header and red border:

<code-example path="docs-style-guide/src/app/not-great.component.ts" header="docs-style-guide/src/app/not-great.component.ts" region="not-great" class="avoid">

</code-example>

{@a code-tabs}

### Code Tabs

Code tabs display code much like `code-examples` with the added advantage of displaying multiple code samples within a tabbed interface.
Each tab displays code using a `code-pane`.

> #### `Code-tabs` attributes

Reviewer 2: If this is a code snippet then it should not be title cased...

```suggestion
#### `code-tabs` attributes
```

* `linenums`: The value can be `true`, `false` or a number indicating the starting line number.
The default is `false`.

> #### Code-pane attributes

Reviewer 2: Should this be a code snippet?

```suggestion
#### `code-pane` attributes
```

/* #enddocregion */
```

The documentation generation process erases these comments before displaying them in the documentation viewer, StackBlitz, and sample code downloads.

<div class="alert is-important">

Because JSON does not allow comments, code snippet markup doesn't work in JSON files.
See the section on [JSON files](#json-files) for more information.

</div>

### `#docregion`

Use `#docregion` in source files to mark code for use in `<code-example>` or `<code-tabs>` components.

The `#docregion` comment begins a code snippet region.
Every line of code after that comment belongs in the region until the code fragment processor encounters the end of the file or a closing `#enddocregion`.

> The following `src/main.ts` is a an example of a file with a single `#docregion` at the top of the file.

Author: Yes, this is as intended.

  </code-pane>
  <code-pane header="app.component.ts (class-skeleton #docregion)" path="docs-style-guide/src/app/app.component.ts" region="class-skeleton">

  </code-pane>

</code-tabs>

The above example also demonstrates that one `#docregion` or `#enddocregion` comment can specify two region names, which is a convenient way to start or stop multiple regions on the same code line.
Alternatively, you could put these comments on separate lines as in the following example.

<code-example header="src/app/app.component.ts">

// #docplaster
...
// #docregion class
// #docregion class-skeleton
export class AppComponent {
// #enddocregion class-skeleton
  title = 'Authors Style Guide Sample';
  heroes = HEROES;
  selectedHero: Hero;

  onSelect(hero: Hero): void {
    this.selectedHero = hero;
  }
// #docregion class-skeleton
}
// #enddocregion class
// #enddocregion class-skeleton

</code-example>

### JSON files

The `<code-example>` component cannot display portions of a JSON file because JSON forbids comments.
However, you can display an entire JSON file by referencing it in the `<code-example>` `src` attribute.

> For large JSON files, you could copy the nodes-of-interest into markdown backticks, but as it's easy to mistakenly create invalid JSON that way, consider creating a JSON partial file with the fragment you want to display.

Author: Fixed Markdown, but omitting code because JSON is notation rather than code, which implies logic.

<live-example title="Live Example with title"></live-example>

You can achieve the same effect by putting the label between the `<live-example>` tags:

```html
<live-example>Live example with content label</live-example>
```

The browser renders the following:

<live-example>Live example with content label</live-example>

### Live example from another guide

To link to an example in a folder where the name is not the same as the current guide page, set the `name` attribute to the name of that folder.

For example, to include the [Router](guide/router) guide example in this style guide, set the `name` attribute to `router`, that is, the name of the folder where that example resides.

```html
<live-example name="router">Live example from the Router guide</live-example>
```

<live-example name="router">Live example from the Router guide</live-example>

### Live Example without download

To omit the download link, add the `noDownload` attribute.

```html
<live-example noDownload>Just the Stackblitz</live-example>
```

The browser renders the following:

<live-example noDownload>Just the Stackblitz</live-example>

### Live Example with download-only

> To omit the live Stackblitz link and only link to the download, add the `downloadOnly` attribute.

Author: Changed. They should all be StackBlitz.

You can embed the StackBlitz within the guide page by adding the `embedded` attribute.

For performance reasons, StackBlitz does not start right away.
Instead, the `<live-example>` component renders an image.
Clicking the image starts the process of launching the embedded StackBlitz within an `<iframe>`.

The following is an embedded `<live-example>` for this guide.

```html
<live-example embedded></live-example>
```

The browser renders the following `<iframe>` and a `<p>` with a link to download the example:

<live-example embedded></live-example>

{@a anchors}

## Anchors

Every section header is also an anchor point.
Another guide page could add a link to this "Anchors" section with the following:

```html
<div class="alert is-helpful">

See the ["Anchors"](guide/docs-style-guide#anchors "Style Guide&mdash;Anchors") section for details.

</div>
```

The browser renders the following:

<div class="alert is-helpful">

See the ["Anchors"](guide/docs-style-guide#anchors "Style Guide&mdash;Anchors") section for details.

</div>

Notice that the above example includes a title of "Style Guide&mdash;Anchors".
Use titles on anchors to create tooltips and improve UX.

> When navigating within a page, you can omit the page URL when specifying the link that [scrolls up](#anchors "Anchors") to the beginning of this section as in the following:

Reviewer 3: ```suggestion
When navigating within a page, you can omit the page URL when specifying the link that [scrolls up](#anchors "Anchors") to the beginning of this section, as in the following:
```

```html
... the link that [scrolls up](#anchors "Anchors") to ...
```

{@a section-anchors}

### Section header anchors

While the documentation generator automatically creates anchors for headers based on the header wording, titles can change, which can potentially break any links to that section.

> To mitigate link breakage, add a custom anchor explicitly, just above the heading or text to which it applies, using the special `{@a name}` syntax as follows.

Reviewer 3: ```suggestion
To mitigate link breakage, add a custom anchor explicitly, just above the heading or text to which it applies, using the special `{@a name}` syntax as follows:
```

### Live Example with download-only

To omit the live Stackblitz link and only link to the download, add the `downloadOnly` attribute.

```html
<live-example downloadOnly>Download only</live-example>
```

The browser renders the following:

<live-example downloadOnly>Download only</live-example>

### Embedded live example

By default, a live example link opens a StackBlitz example in a separate browser tab.
You can embed the StackBlitz within the guide page by adding the `embedded` attribute.

For performance reasons, StackBlitz does not start right away.
Instead, the `<live-example>` component renders an image.
Clicking the image starts the process of launching the embedded StackBlitz within an `<iframe>`.

The following is an embedded `<live-example>` for this guide.

```html
<live-example embedded></live-example>
```

The browser renders the following `<iframe>` and a `<p>` with a link to download the example:

<live-example embedded></live-example>

{@a anchors}

## Anchors

Every section header tag is also an anchor point.
Another guide page could add a link to this "Anchors" section with the following:

```html
<div class="alert is-helpful">

> See the ["Anchors"](guide/docs-style-guide#anchors "Style Guide&mdash;Anchors") section for details.

Reviewer 4: This would be the place to recommend tooltips on anchors as well, for clarity and accessibility.

// #docregion class-skeleton
}
// #enddocregion class, class-skeleton

</code-example>

> The `#docplaster` at the top of the file tells the processor to join the fragments into a single snippet without anything in between.

Reviewer 2: ```suggestion
The `#docplaster` marker tells the processor what text string to use (the "plaster") to join each of the fragments into a single snippet. You place the "plaster" text on the same line. For example: `#docplaster ---` would use `---` as the "plaster" text. In the case of the previous file the "plaster" text is empty so there will be nothing in between each fragment.
```

Author: Oh thank you, @petebacondarwin! I didn't know this. Added. 😊

  </code-pane>

</code-tabs>

The above example also demonstrates that one `#docregion`/`#enddocregion` comment can specify two region names, which is a convenient way to start (or stop) multiple regions on the same code line.
Alternatively, you could put these comments on separate lines as in the following example.

<code-example header="src/app/app.component.ts">

// #docplaster
...
// #docregion class
// #docregion class-skeleton
export class AppComponent {
// #enddocregion class-skeleton
  title = 'Authors Style Guide Sample';
  heroes = HEROES;
  selectedHero: Hero;

  onSelect(hero: Hero): void {
    this.selectedHero = hero;
  }
// #docregion class-skeleton
}
// #enddocregion class
// #enddocregion class-skeleton

</code-example>

### JSON files

The `<code-example>` component cannot display portions of a JSON file because JSON forbids comments.
However, you can display an entire JSON file by referencing it in the `<code-example>` `src` attribute.

For large JSON files, you could copy the nodes-of-interest into markdown backticks, but as it's easy to mistakenly create invalid JSON that way, consider creating a JSON partial file with the fragment you want to display.

> You can't test a partial file nor use it in the application, but at least your editor can confirm that it is syntactically correct.

Reviewer 2: And also you can store the partial file next to the original, so it is more likely that the author will remember to keep the two in sync.

Instead, the `<live-example>` component renders an image.
Clicking the image starts the process of launching the embedded StackBlitz within an `<iframe>`.

The following is an embedded `<live-example>` for this guide.

```html
<live-example embedded></live-example>
```

The browser renders the following `<iframe>` and a `<p>` with a link to download the example:

<live-example embedded></live-example>

{@a anchors}

## Anchors

Every section header tag is also an anchor point.
Another guide page could add a link to this "Anchors" section with the following:

```html
<div class="alert is-helpful">

See the ["Anchors"](guide/docs-style-guide#anchors "Style Guide&mdash;Anchors") section for details.

</div>
```

The browser renders the following:

<div class="alert is-helpful">

See the ["Anchors"](guide/docs-style-guide#anchors "Style Guide&mdash;Anchors") section for details.

</div>

When navigating within a page, you can omit the page URL when specifying the link that [scrolls up](#anchors "Anchors") to the beginning of this section as in the following:

```html
... the link that [scrolls up](#anchors "Anchors") to ...
```

{@a long-anchors}

> ### Long section header anchors

Author: Great suggestion. I've re-worked that section to be about explicit anchors and dropped the info about length.

</div>

## Alerts and callouts

Alerts and callouts present warnings, extra detail, or references to related topics.

An alert or callout should not contain anything essential to understanding the main content.
Instructions or tutorial steps should be in the main body of a guide rather than in a subsection.

### Alerts

Alerts draw attention to short, important points.
For multi-line content, see [callouts](#callouts "callouts").

<div class="alert is-helpful">

  See the [live examples](guide/docs-style-guide#live-examples "Live examples") section for more information.

</div>

Note that at least one blank line must follow both the opening and closing `<div>` tags.
A blank line before the closing `</div>` is conventional but not required.

```html
<div class="alert is-helpful">

  See the [live examples](guide/docs-style-guide#live-examples "Live examples") section for more information.

</div>
```

There are three different levels for styling the alerts according to the importance of the content.

> Use the following the `alert` class along with the appropriate `is-helpful`, `is-important`, or `is-critical` CSS class as follows.

Reviewer 3: ```suggestion
Use the following `alert` class along with the appropriate `is-helpful`, `is-important`, or `is-critical` CSS class, as follows:
```

<div class="callout is-helpful">

<header>A helpful or informational point</header>

  **A helpful note**. Use a helpful callout for information requiring explanation.
  Callouts are typically multi-line notes.
  They can also contain code snippets.

</div>
```

The browser renders the three styles as follows:

<div class="callout is-helpful">

  <header>A helpful or informational point</header>

  **A helpful note**. Use a helpful callout for information requiring explanation.
  Callouts are typically multi-line notes.
  They can also contain code snippets.

</div>

<div class="callout is-important">

  <header>An important point</header>

  **An important note**. Use an important callout for significant information requiring explanation.
  Callouts are typically multi-line notes.
  They can also contain code snippets.

</div>

<div class="callout is-critical">

  <header>A critical point</header>

  **A critical note**. Use a critical callout for compelling information requiring explanation.
  Callouts are typically multi-line notes.
  They can also contain code snippets.

</div>

When using callouts, consider the following points:

* The callout header text style is uppercase.

> * The header does not render in the page table of contents.

Reviewer 3: ```suggestion
* The header does not render in the table of contents.
```

* You can write the callout body in markdown.
* A blank line separates the `<header>` tag from the markdown content.

> * Avoid using an `<h2>`, `<h3>`, `<h4>`, `<h5>`, or `<h6>` as the CSS for callouts styles the `<header>` element.

Reviewer 3: ```suggestion
* Avoid using an `<h2>`, `<h3>`, `<h4>`, `<h5>`, or `<h6>`, as the CSS for callouts styles the `<header>` element.
```

</div>

<div class="alert is-critical">

A critical alert.

</div>

### Callouts

Callouts, like alerts, highlight important points.

> Use a callout when you need a header and multi-line content.

Reviewer 4: We could mention that `<header>` is the same as using h4 here -- neither appears in the page TOC. 
Possibly "Avoid using h2 or h3"?

</div>
```

## Tables

Use HTML tables to present tabular data.

<table>
  <tr>
    <th>Header Type</th>
    <th>Markdown</th>
    <th>Use</th>
  </tr>
  <tr>
    <td><code>&lt;h1&gt;</code></td>
    <td>#</td>
    <td>Title of page</td>
  </tr>
  <tr>
    <td><code>&lt;h2&gt;</code></td>
    <td>##</td>
    <!-- can use markdown too; remember blank lines -->
    <td>

      **Second level headers**

    </td>
  </tr>
  <tr>
    <td><code>&lt;h3&gt;</code></td>
    <td>###</td>
    <td>

      **Third level headers**

    </td>
  </tr>
</table>

> The following is the markup for this table.

Reviewer 3: ```suggestion
The following is the markup for this table:
```

```html

<table>
  <tr>
    <th>Header Type</th>
    <th>Markdown</th>
    <th>Use</th>
  </tr>
  <tr>
    <td><code>&lt;h1&gt;</code></td>
    <td>#</td>
    <td>Title of page</td>
  </tr>
  <tr>
    <td><code>&lt;h2&gt;</code></td>
    <td>##</td>
    <!-- can use markdown too; remember blank lines -->
    <td>

      **Second level headers**

    </td>
  </tr>
  <tr>
    <td><code>&lt;h3&gt;</code></td>
    <td>###</td>
    <td>

      **Third level headers**

    </td>
  </tr>
</table>

```

## Images

Store images in the `content/images/guide` directory in a folder with the **same name** as the guide page.
Because Angular documentation generation copies these images to `generated/images/guide/your-guide-directory`, set the image `src` attribute to the runtime location of `generated/images/guide/your-guide-directory`.

For example, images for this "Angular documentation style guide" are in the `content/images/guide/docs-style-guide` folder, but the `src` attribute specifies the `generated` location.

> The following is the `src` attribute for the "flying hero" image belonging to this guide.

Reviewer 3: ```suggestion
The following is the `src` attribute for the "flying hero" image belonging to this guide:
```

```html
src="generated/images/guide/docs-style-guide/flying-hero.png"
```

<div class="callout is-important">

  <header>Use the HTML <code>&lt;img&gt;</code> tag</header>

  Specify images using the `<img>` tag.

>   **Do not use the markdown image syntax, \!\[\.\.\.\]\(\.\.\.\).**

Reviewer 3: ```suggestion
  **Do not use the Markdown image syntax, \!\[\.\.\.\]\(\.\.\.\).**
```

  Nest the `<img>` tag within a `<div class="lightbox">` tag, which styles the image according to the documentation standard.

  ```html
  <div class="lightbox">
    <img src="generated/images/guide/docs-style-guide/flying-hero.png" alt="flying hero">
  </div>
  ```

  Note that the HTML `<img>` element does not have a closing tag.

  The browser renders the following:

  <div class="lightbox">
    <img src="generated/images/guide/docs-style-guide/flying-hero.png" alt="flying hero">
  </div>

</div>

### Image captions and figure captions

A caption appears underneath the image as a concise and comprehensive summary of the image.
Captions are optional unless you are using numbered figures, such as Figure 1, Figure 2, and so on.
If you are using numbered figures in a page, follow the guidelines in [Figure captions](https://developers.google.com/style/images#figure-captions) in the Google Developer Documentation Style Guide.

### Image dimensions

The doc generator reads the image dimensions from an image file and adds `width` and `height` attributes to the `<img>` tag automatically.

To control the size of the image, supply your own `width` and `height` attributes.

> Here's the "flying hero" markup with a 200px width.

Reviewer 3: ```suggestion
Here's the "flying hero" markup with a 200px width:
```

```html

<div class="lightbox">
  <img src="generated/images/guide/docs-style-guide/flying-hero.png"
    alt="flying Angular hero"
    width="200">
</div>
```

The browser renders the following:

<div class="lightbox">

  <img src="generated/images/guide/docs-style-guide/flying-hero.png"
    alt="flying Angular hero"
    width="200">

</div>

### Wide images

To prevent images overflowing their viewports, **use image widths under 700px**.

> To display a larger image, provide a link to the actual image that the user can click to see the full size image separately as in this example of `source-map-explorer` output from the "Ahead-of-time Compilation" guide:

Reviewer 3: ```suggestion
To display a larger image, provide a link to the actual image that the user can click to see the full size image separately, as in this example of `source-map-explorer` output from the "Ahead-of-time Compilation" guide:
```

<a href="generated/images/guide/docs-style-guide/toh-pt6-bundle.png" title="Click to view larger image">
  <div class="lightbox">
    <img src="generated/images/guide/docs-style-guide/toh-pt6-bundle-700w.png" alt="toh-pt6-bundle" width="300px">
  </div>
</a>

> The following is the HTML for creating a link to the image.

Reviewer 3: ```suggestion
The following is the HTML for creating a link to the image:
```

</div>
```

## Tables

Use HTML tables to present tabular data.

> <!--

Reviewer 4: I thought the recommendation was to use markdown tables when possible. At least, it was at one time. They are easier to read in source.

Author: I wasn't sure one way or the other, so took the question to the wider team. For now we'll stick with HTML tables because they are flexible and style more easily.

For example, images for this "Angular documentation style guide" are in the `content/images/guide/docs-style-guide` folder, but the `src` attribute specifies the `generated` location.

The following is the `src` attribute for the "flying hero" image belonging to this guide.

```
src="generated/images/guide/docs-style-guide/flying-hero.png"
```

<div class="callout is-important">

  <header>Use the HTML <code>&lt;img&gt;</code> tag</header>

  Specify images using the `<img>` tag.
  **Do not use the markdown image syntax, \!\[\.\.\.\]\(\.\.\.\).**

  For accessibility, always set the `alt` attribute with a meaningful description of the image.

  Nest the `<img>` tag within a `<div class="lightbox">` tag, which styles the image according to the docs standard.

  ```html
  <div class="lightbox">
    <img src="generated/images/guide/docs-style-guide/flying-hero.png" alt="flying hero">
  </div>
  ```

  Note that the HTML `<img>` element does not have a closing tag.

  The browser renders the following:

  <div class="lightbox">
    <img src="generated/images/guide/docs-style-guide/flying-hero.png" alt="flying hero">
  </div>

</div>

> ### Image dimensions

Reviewer: Possibly add this short section.

> ### Image captions and figure captions
> 
> A caption appears underneath the image as a concise and comprehensive summary of the image. Captions are optional unless you are using numbered figures (such as Figure 1, Figure 2, etc.). If you are using numbered figures in a page, follow the guidelines in [Figure captions](https://developers.google.com/style/images#figure-captions) in the Google Developer Documentation Style Guide.
>

  A subsection with **markdown** formatted text.

</div>

## Help with documentation style

For specific language and grammar usage, a word list, style, tone, and formatting recommendations, see the [Google Developer Documentation Style Guide](https://developers.google.com/style).

> If you have any questions that this style guide doesn't answer or would like to discuss documentation styles visit the [Angular repo](https://github.com/angular/angular) and [file a documentation issue](https://github.com/angular/angular/issues/new/choose).

Reviewer 3: ```suggestion
If you have any questions that this style guide doesn't answer or you would like to discuss documentation styles, visit the [Angular repo](https://github.com/angular/angular) and [file a documentation issue](https://github.com/angular/angular/issues/new/choose).
```

  A subsection with **markdown** formatted text.

</div>

## Help with docs styles

> If you have any questions not included in this style guide or would like to discuss docs styles visit the [Angular repo](https://github.com/angular/angular) and [file a docs issue](https://github.com/angular/angular/issues/new/choose).

Reviewer: This would be a good place to put in a general link to the [Google Developer Documentation Style Guide](https://developers.google.com/style) for specific language and grammar usage, a word list, style and tone, and formatting.
